### PR TITLE
feature(gg15): add gumbel softmax

### PR DESCRIPTION
**Pull request by shgqmrf15**
## Description
add gumbel softmax
for more gumbelsoftmax info, you can refer to the paper <https://arxiv.org/abs/1611.01144>
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/torch_utils/network/gumbel_softmax.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        r"""
+        Overview:
+            Initialize the GumbelSoftmax module
+        """
+        super(GumbelSoftmax, self).__init__()
+
+    def sample_gumbel(self, shape, device, eps=1e-20):
+        """Sample from Gumbel(0, 1)"""
+        U = torch.rand(shape)
+        if device == torch.device("cuda"):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gumbel_softmax.py:29]: `U = U.to(device)`

File: ding/torch_utils/network/gumbel_softmax.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            U = U.cuda()
+        return -torch.log(-torch.log(U + eps) + eps)
+
+
+    def gumbel_softmax_sample(self, x: torch.Tensor, temperature):
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gumbel_softmax.py:34]: merge `gumbel_softmax_sample` and  `sample_gumbel` into `forward` method

File: ding/torch_utils/network/gumbel_softmax.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        """ Draw a sample from GumbelSoftmax distribution"""
+        y = x + self.sample_gumbel(x.shape, x.device)
+        return F.softmax(y/temperature, dim=1)
+
+    def forward(self, x: torch.Tensor, temperature=1, hard=False) -> torch.Tensor:
+        r"""
+        Arguments:
+            - x (:obj:`torch.Tensor`): unnormalized log-probs
+            - temperature: non-negative scalar
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gumbel_softmax.py:43]: fill all the comments

File: ding/torch_utils/network/gumbel_softmax.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            U = U.cuda()
+        return -torch.log(-torch.log(U + eps) + eps)
+
+
+    def gumbel_softmax_sample(self, x: torch.Tensor, temperature):
+        """ Draw a sample from GumbelSoftmax distribution"""
+        y = x + self.sample_gumbel(x.shape, x.device)
+        return F.softmax(y/temperature, dim=1)
+
+    def forward(self, x: torch.Tensor, temperature=1, hard=False) -> torch.Tensor:
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gumbel_softmax.py:39]: fill all the python typing lint

File: ding/torch_utils/network/gumbel_softmax.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            - hard: if true return one-hot label
+        Returns:
+            - output (:obj:`torch.Tensor`): sample from GumbelSoftmax distribution
+        Shapes:
+            - x: :math:`(B, N)`, while B is the batch size, N is number of classes
+            - output: :math:`(B, N)`, while B is the batch size, N is number of classes
+        """
+        y = self.gumbel_softmax_sample(x, temperature)
+        if hard:
+            y_hard = torch.zeros(x.shape)
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gumbel_softmax.py:53]: `y_hard = torch.zeros_like(x)`

File: ding/torch_utils/network/gumbel_softmax.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            y_hard[torch.arange(0,x.shape[0]).to(int),y.max(1)[1]] = 1
+            y = (y_hard - y).detach() + y
```
Comment (PaParaZz1) [inline ding/torch_utils/network/gumbel_softmax.py:55]: add some comments for this implementation

File: ding/torch_utils/network/tests/test_gumbel_softmax.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+class TestGumbelSoftmax:
+    def test(self):
+        B = 4
+        N = 10
+        model = GumbelSoftmax()
+        # data case 1
+        for _ in range(N):
+            data = torch.rand((4,10))
+            data = torch.log(data)
+            GumbelSoftmaxDistri = model(data, hard=True)
```
Comment (PaParaZz1) [inline ding/torch_utils/network/tests/test_gumbel_softmax.py:17]: add test case for `hard=False`, you can refer to [this](https://github.com/opendilab/DI-engine/blob/main/ding/rl_utils/tests/test_ppo.py#L25)

Comment (PaParaZz1) [inline ding/torch_utils/network/tests/test_gumbel_softmax.py:17]: modify variable name `GumbelSoftmaxDistri`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]
